_posts/2025-12-08-datetime-library.markdown

```diff
@@ -0,0 +1,575 @@
+---
+layout: post
+title:  "A Verified DateTime Library for Dafny"
+author: Jerry Zhou, Yuyang Liu, Jules Wang
+date:   2025-12-08 09:00:00 +0000
+categories: standard-libraries
+---
+
+# DateTime Library
+
+Date and time handling is one of the most error-prone areas in software development.
+
+From leap year calculations to timezone transitions, temporal data presents edge cases that can corrupt application logic. This becomes critical in verification languages like Dafny, where correctness is mathematically proven.
+
+We built `Std.LocalDateTime` and `Std.ZonedDateTime`, comprehensive modules providing timezone-agnostic and timezone-aware date-time operations with full formal verification. This post covers:
+
+* design principles behind safe temporal arithmetic,
+* validation and bounded integer types,
+* timezone handling and DST resolution,
+* parsing and formatting with formal contracts,
+* epoch-based arithmetic for reliable calculations,
+* verification challenges we solved.
+
+
+## Design principles
+
+Our design balances mathematical precision with real-world complexity:
+
+1. **Safety through types:** Bounded integer types (`int32`, `uint8`, `uint16`) prevent overflow while maintaining verification guarantees.
+
+2. **Explicit validation:** Every instance must satisfy validity predicates, checking leap years, month boundaries, time constraints, and (for zoned times) offset bounds.
+
+3. **Immutability:** Pure functions like `WithYear`, `WithMonth` return new validated instances instead of mutation.
+
+4. **Epoch-based arithmetic:** Date arithmetic converts to epoch milliseconds, performs calculation, then converts back—avoiding calendar complexity.
+
+5. **Parse-don't-validate:** Parsing functions return `Result<T, string>`, forcing explicit error handling.
```

Comment:
"Parse-don't-validate" seems to conflict with the "Parsing with validation" header below. I think it's more correct to say parsing does involve "validation" in the sense of runtime checks, which makes sense.